### PR TITLE
cyw43: Add Control method to add multicast HW address

### PR DESCRIPTION
**Pull request by sammko**
Notably, a method to remove addresses is missing. I can add this if wanted but I ran out of time for now.

In reference to #2176.

### CODE REVIEW

File: cyw43/src/control.rs [hunk 1]
```diff
@@ -316,6 +322,50 @@ impl<'a> Control<'a> {
...
+        free[0] = address;
+        let n = n + 1;
+        buf[..4].copy_from_slice(&(n as u32).to_le_bytes());
+
+        self.set_iovar_v::<80>("mcast_list", &buf).await;
+        Ok(n)
+    }
+
+    /// Retrieve the list of configured multicast hardware addresses.
+    pub async fn list_mulistcast_addresses<'b>(&mut self, buf: &'b mut [u8; 64]) -> &'b [[u8; 6]] {
```
Comment (sammko) [inline cyw43/src/control.rs:360]: I'm not sure about this pattern of using a caller-provided buffer and returning a typed slice inside it. However, without `heapless::Vec` or something equivalent, I don't see a better solution.

File: cyw43/src/control.rs [hunk 2]
```diff
@@ -16,6 +16,12 @@ pub struct Error {
     pub status: u32,
 }
 
+#[derive(Debug)]
+pub enum AddMulticastAddressError {
+    NotMulticast,
+    NoFreeSlots,
+}
+
```
Comment (sammko) [inline cyw43/src/control.rs:24]: Is this error handling strategy appropriate in this library? I.e. a separate error type for this use-case, not mixed with the other one above. I know it's a matter of convention, let me know what is best here.

Comment (Dirbaio) [reply cyw43/src/control.rs:24]: :+1:

File: cyw43/src/control.rs (hunk 1 above)
Comment (Dirbaio) [reply cyw43/src/control.rs:360]: it's a bit strange yes, especially changing the type. in these cases it's more common to take a buffer and return how many slots have been filled:

`pub async fn list_mulistcast_addresses(&mut self, addresses: &mut [[u8; 6]; 10]) -> usize`

File: cyw43/src/lib.rs [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #![no_std]
 #![no_main]
-#![feature(async_fn_in_trait, type_alias_impl_trait, concat_bytes)]
+#![feature(async_fn_in_trait, type_alias_impl_trait, concat_bytes, slice_as_chunks)]
```
Comment (Dirbaio) [inline cyw43/src/lib.rs:3]: We shouldn't add more nightly features, the goal is for Embassy to work on stable when AFIT and TAIT are stable.

File: cyw43/src/control.rs (hunk 1 above)
Comment (sammko) [reply cyw43/src/control.rs:360]: Thing is I need spaces in front of the addresses for the 4 bytes length. Maybe we can just pull in `heapless`, as it's a dependency of `embassy-net` anyway? I think that would offer the nicest API.

File: cyw43/src/lib.rs (hunk 3 above)
Comment (sammko) [reply cyw43/src/lib.rs:3]: Noted, I'll do it without. 👍

File: cyw43/src/control.rs (hunk 1 above)
Comment (Dirbaio) [reply cyw43/src/control.rs:360]: hmm perhaps read to a stack buffer and copy the interesting part to the user buffer then? it uses 2x the memory, but you need that anyway if you return a heapless::Vec instead. Also, the compiler tends to be dumb about returning big things by value, also causing multiple copies of it on the stack.

**Comment by sammko** [comment]
Should all be addressed now.

**Review by Dirbaio — APPROVED** [review]
thank you!

**Pull request merged**